### ISSUE TITLE
aws-sdk-ios generated machine learning string to sign does not match with the signature calculated

### ISSUE THREAD

**Issue by vamcyleonardo** [aws/aws-sdk-ios#331]
Hi,
I am trying to predict using machine learning and getting the following error.
```
message=The request signature we calculated does not match the signature you provided. Check your AWS Secret Access Key and signing method. 
The String-to-Sign should have been
'AWS4-HMAC-SHA256
20160315T103519Z
20160315/us-east-1/machinelearning/aws4_request
853c2137861a8b1903416500e41f1bcb32783c755b35be6515aaaeda52a2ac6c'
```
However, in the request i am performing, the first part .GetMLModel as shown below to get connected to end point url was successful and the later part .Predict fails with the above error. I am guessing the extra '/' after POST in .Predict canonical request is the culprit here. Am i right? If yes, how do i get to fix this? If no, what is the possible error? Thank you.
```
AWS4 Canonical Request: [POST/
content-type:application/x-amz-json-1.1
host:machinelearning.us-east-1.amazonaws.com
user-agent:aws-sdk-iOS/2.3.5 iPhone-OS/9.2 en_US
x-amz-date:20160316T054451Z
x-amz-target:AmazonML_20141212.GetMLModel
content-type;host;user-agent;x-amz-date;x-amz-target
24f129cce16214ecf6584d72aa625c24390f38674c5c37eb041e9bd044b94ed1]
```
```
AWS4 Canonical Request: [POST//
content-type:application/x-amz-json-1.1
host:realtime.machinelearning.us-east-1.amazonaws.com,realtime.machinelearning.us-east-1.amazonaws.com
user-agent:aws-sdk-iOS/2.3.5 iPhone-OS/9.2 en_US
x-amz-date:20160316T054458Z
x-amz-target:AmazonML_20141212.Predict
content-type;host;user-agent;x-amz-date;x-amz-target
da2d7f28d79e55082d66964fce4882a87657975222a8dfb5fb7c0d014d274de0]
```

**Comment by yosuke-matsuda** [comment]
Could you share a code snippet that we can use to reproduce the issue? Thanks.

**Comment by vamcyleonardo** [comment]
Hi Yosuke,
This is the code snippet.

//App Delegate.m

#import "AppDelegate.h"
#import "AWSCore.h"
#import "AWSMachineLearning.h"

@interface AppDelegate ()

@end

@implementation AppDelegate


- (BOOL)application:(UIApplication *)application didFinishLaunchingWithOptions:(NSDictionary *)launchOptions {
    AWSStaticCredentialsProvider *credentialsProvider = [[AWSStaticCredentialsProvider alloc]
                                                         initWithAccessKey:@"YOUR_ACCESS_KEY_HERE"
                                                         secretKey:@"YOUR_SECRET_KEY_HERE"];
    AWSServiceConfiguration *serviceConfiguration = [[AWSServiceConfiguration alloc]
                                                     initWithRegion:AWSRegionUSEast1
                                                    credentialsProvider:credentialsProvider];
    [AWSServiceManager defaultServiceManager].defaultServiceConfiguration = serviceConfiguration;

    return YES;
}



import "ViewController.h"


#import "AWSMachineLearning.h"


@interface ViewController ()


@end


// Cache real-time endpoint

NSString *endpoint;


// Amazon Machine Learning Client

AWSMachineLearning *machineLearning;


@implementation ViewController


- (void)viewDidLoad {

    [super viewDidLoad];

    

    machineLearning = [AWSMachineLearning defaultMachineLearning];

    

    // Specify your ML Model

    NSString *MLModelId = @"YOUR-ML-MODEL-ID";

    

    AWSMachineLearningGetMLModelInput *getMLModelInput = [AWSMachineLearningGetMLModelInput new];

    getMLModelInput.MLModelId = MLModelId;

    // Call Get ML Model

    [[[machineLearning getMLModel:getMLModelInput] continueWithSuccessBlock:^id(AWSTask *task) {

        AWSMachineLearningGetMLModelOutput *getMLModelOutput = task.result;

        

        if (getMLModelOutput.status != AWSMachineLearningEntityStatusCompleted) {

            NSLog(@"ML Model is not completed");

            return nil;

        }

        if (getMLModelOutput.endpointInfo.endpointStatus != AWSMachineLearningRealtimeEndpointStatusReady) {

            NSLog(@"Real-time endpoint is not ready");

            return nil;

        }

        endpoint = getMLModelOutput.endpointInfo.endpointUrl;

        

        // Since model is complete and real-time endpoint is ready, call predict

        return [self predict:MLModelId withRecord: @{}];


    }] continueWithBlock:^id(AWSTask *task) {

        if (task.error) {

            NSLog(@"Error %@", task.error);

        }

        if (task.exception) {

            NSLog(@"Exception %@", task.exception);

        }

        if (task.result) {

            AWSMachineLearningPredictOutput *predictOutput = task.result;

            NSLog(@"Prediction: %@", predictOutput.prediction);

        }

        return nil;

    }];

}


- (AWSTask *) predict:(NSString *)mlModelId withRecord:(NSDictionary *)record {

    AWSMachineLearningPredictInput *predictInput = [AWSMachineLearningPredictInput new];

    predictInput.predictEndpoint = endpoint;

    predictInput.MLModelId = mlModelId;

    predictInput.record = record;

    return [machineLearning predict:predictInput];
}

**Comment by yosuke-matsuda** [comment]
Can you change the code to construct `AWSMachineLearningPredictInput` to something like the following and see if you continue seeing the error?
```
AWSMachineLearningGetMLModelOutput *getMLModelOutput = task.result;

AWSMachineLearningPredictInput *predictInput = [AWSMachineLearningPredictInput new];
NSString *endpointUrl = getMLModelOutput.endpointInfo.endpointUrl;
if ([endpointUrl hasSuffix:@"/"]) {
    endpointUrl = [endpointUrl substringToIndex:[getMLModelOutput.endpointInfo.endpointUrl length] -1];
}
```
It is possible that the service started returning an extra `/` at the end of `endpointUrl`. By removing this, it may resolve the issue.

Thanks,

**Comment by vamcyleonardo** [comment]
Sorry for the late response yosuke,
Yes the solution you provided worked. Thank you. But it actually worked for me without this fix initially, before this bug started occurring. Later on it would never work without removing the '\'.

**Comment by yosuke-matsuda** [comment]
The service behavior changed recently to include the trailing `/`. That's why it used to work but suddenly started failing.

The [API Reference](http://docs.aws.amazon.com/AWSiOSSDK/latest/Classes/AWSMachineLearning.html#//api/name/defaultMachineLearning) may be helpful (although the Swift code snippet label says "Objective-C").

**Comment by yosuke-matsuda** [comment]
We've released the AWS Mobile SDK for iOS 2.4.0, which addresses the issue. Thanks.
